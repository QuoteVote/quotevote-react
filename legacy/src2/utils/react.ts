import React, { 
  useState, 
  useEffect, 
  useRef, 
  ReactNode, 
  ChangeEvent, 
  KeyboardEvent,
  FormEvent,
  MouseEvent,
  Dispatch,
  SetStateAction
} from 'react';

export type { 
  ReactNode,
  ChangeEvent,
  KeyboardEvent,
  FormEvent,
  MouseEvent,
  Dispatch,
  SetStateAction
};

export { 
  React as default,
  useState,
  useEffect,
  useRef
};

// Add type declarations for common event handlers
export type InputChangeEvent = ChangeEvent<HTMLInputElement>;
export type TextAreaChangeEvent = ChangeEvent<HTMLTextAreaElement>;
export type ButtonClickEvent = MouseEvent<HTMLButtonElement>;
export type SetState<T> = Dispatch<SetStateAction<T>>;

// Add type declarations for common component props
export interface BaseProps {
  className?: string;
  style?: React.CSSProperties;
  children?: ReactNode;
}

export interface InputProps extends BaseProps {
  value: string;
  onChange: (value: string) => void;
  placeholder?: string;
  disabled?: boolean;
} 