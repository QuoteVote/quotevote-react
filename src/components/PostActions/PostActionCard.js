import React, { useEffect, useState } from 'react'
import {
  Card, CardActions, CardContent, IconButton, Typography,
} from '@material-ui/core'
import { InsertEmoticon, InsertLink } from '@material-ui/icons'
import { makeStyles } from '@material-ui/core/styles'
import SvgIcon from '@material-ui/core/SvgIcon'
import PropTypes from 'prop-types'
import { useDispatch, useSelector } from 'react-redux'
import { get } from 'lodash'
import copy from 'clipboard-copy'
import SweetAlert from 'react-bootstrap-sweetalert'
import { useHistory } from 'react-router-dom'
import AvatarDisplay from '../Avatar'
import { parseCommentDate } from '../../utils/momentUtils'
import { SET_FOCUSED_COMMENT, SET_SHARED_COMMENT } from '../../store/ui'
import { ReactComponent as DislikeIcon } from '../../assets/svg/Dislike.svg'
import { ReactComponent as LikeIcon } from '../../assets/svg/Like.svg'
import { ReactComponent as QuoteIcon } from '../../assets/svg/Quote.svg'
import { ReactComponent as CommentIcon } from '../../assets/svg/Comment.svg'
import buttonStyle from '../../assets/jss/material-dashboard-pro-react/components/buttonStyle'
import PostChatMessage from '../PostChat/PostChatMessage'

const useStyles = makeStyles((theme) => ({
  content: {
    marginLeft: 10,
    marginRight: 40,
    marginTop: 10,
    marginBottom: -20,
    fontSize: 16,
  },
  expand: {
    marginLeft: 'auto',
  },
  created: {
    verticalAlign: 'middle',
    marginTop: 20,
    marginRight: 10,
  },
  root: {
    backgroundColor: theme.palette.background.paper,
    width: '100%',
  },
  selectedRoot: {
    backgroundColor: '#f1e8c1',
    width: '100%',
  },
  ...buttonStyle,
  date: {
    color: '#949292',
  },
}))

function PostActionCard({ postAction, postUrl, selected }) {
  const history = useHistory()
  const classes = useStyles()
  const dispatch = useDispatch()
  const {
    user, content, created, _id,
  } = postAction
  const { username, avatar, name } = user
  const parsedDate = parseCommentDate(created)
  const voteType = get(postAction, 'type')
  const quote = get(postAction, 'quote')
  const sharedComment = useSelector((state) => state.ui.sharedComment)

  const baseUrl = window.location.origin
  const handleCopy = async () => {
    await copy(`${baseUrl}${postUrl}/comment#${_id}`)
    setOpen(true)
  }
  const [open, setOpen] = useState(false)
  const hideAlert = () => {
    setOpen(false)
  }

  let postContent = content
  let svgIcon = CommentIcon
  let voteTags = ''

  const handleRedirectToProfile = () => {
    history.push(`/hhsb/Profile/${username}`)
  }

  if (voteType) {
    const isUpvote = voteType === 'up'
    const defaultTag = isUpvote ? '#agree' : '#disagree'
    svgIcon = isUpvote ? LikeIcon : DislikeIcon
    voteTags = get(postAction, 'tags') || defaultTag
  }

  if (quote) {
    postContent = quote.length ? quote : 'Quoted this post.'
    svgIcon = QuoteIcon
  }

  useEffect(() => {
    if (selected) {
      dispatch(SET_SHARED_COMMENT(postAction))
      dispatch(SET_FOCUSED_COMMENT(postAction))
    }
  }, [postAction, selected, dispatch])

  if (postAction.text) {
    return (
      <PostChatMessage message={postAction} key={postAction._id} />
    )
  }

  return (
    <Card
      onMouseEnter={() => dispatch(SET_FOCUSED_COMMENT(postAction))}
      onMouseLeave={() => dispatch(SET_FOCUSED_COMMENT(sharedComment))}
      className={selected ? classes.selectedRoot : classes.root}
    >
      {!voteType && (
        <CardContent
          className={classes.content}
        >
          <p>
            {postContent}
          </p>
        </CardContent>
      )}
      <CardActions disableSpacing>
        <IconButton
          onClick={() => handleRedirectToProfile()}
        >
          <AvatarDisplay height={20} width={20} {...avatar} />
        </IconButton>
        <Typography display="inline">
          {name}
          {' '}
          <span className={classes.date}>{parsedDate}</span>
        </Typography>
        <SvgIcon
          component={svgIcon}
          fontSize="large"
          viewBox="-10 -10 50 50"
          htmlColor="black"
        />
        <Typography display="inline">{voteTags}</Typography>
        <IconButton className={classes.expand}>
          <InsertEmoticon />
        </IconButton>
        <IconButton onClick={handleCopy}>
          <InsertLink />
        </IconButton>
      </CardActions>
      {open && (
        <SweetAlert
          confirmBtnCssClass={`${classes.button} ${classes.success}`}
          success
          onConfirm={hideAlert}
          onCancel={hideAlert}
          title="Comment URL copied!"
          timeout={1000}
        />
      )}
    </Card>
  )
}

PostActionCard.propTypes = {
  postAction: PropTypes.object.isRequired,
  postUrl: PropTypes.string,
  selected: PropTypes.bool,
}

export default PostActionCard
