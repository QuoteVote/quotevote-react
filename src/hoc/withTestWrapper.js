import React from 'react'
import { PersistGate } from 'redux-persist/integration/react'
import { ApolloProvider } from '@apollo/react-hooks'
import { Provider } from 'react-redux'
import { createTheme, ThemeProvider } from '@material-ui/core/styles'
import { Router } from 'react-router-dom'
import { createBrowserHistory } from 'history'
import store, { persistor } from '../store/store'
import client from '../config/apollo'
import customTheme from '../theme'
const hist = createBrowserHistory()
const theme = createTheme(customTheme)

const withTestWrapper = (Component) => (props) => (
  <ApolloProvider client={client}>
    <Provider store={store}>
      <PersistGate loading={null} persistor={persistor}>
        <ThemeProvider theme={theme}>
          <Router history={hist}>
            <Component {...props} />
          </Router>
        </ThemeProvider>
      </PersistGate>
    </Provider>
  </ApolloProvider>
)
export default withTestWrapper
