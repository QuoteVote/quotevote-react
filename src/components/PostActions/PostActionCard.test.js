import React from 'react'
import { render } from '@testing-library/react'

// Component being tested
import PostActionCard from './PostActionCard'
import withTestWrapper from '../../hoc/withTestWrapper'

const PostActionCardWrapper = withTestWrapper(PostActionCard)

describe('PostActionCard test -', () => {
  it('renders correctly', () => {
    const { container } = render(
      <PostActionCardWrapper />
    )
    expect(container.firstChild).toMatchSnapshot()
  })
})
