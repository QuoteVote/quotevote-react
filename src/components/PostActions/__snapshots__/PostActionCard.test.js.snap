// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostActionCard test - renders correctly 1`] = `null`;
